### ISSUE TITLE
scons fails to build for stm32f0xx devices.

### ISSUE THREAD

**Issue by lmoesch** [roboterclubaachen/xpcc#345]
```
[xpcc\build\stm32f0_discovery\blink\main.o] UnicodeDecodeError : 'charmap' codec can't decode byte 0x81 in position 42: character maps to <undefined>
scons: building terminated because of errors.
```

Get this error for the stm32f0_discovery/blink example, aswell as my regular project code, which builds flawlessly for e.g. the stm32f746g_discovery or the f4 nucleo boards.

**Comment by salkinium** [comment]
Position 42 in `stm32f0_discovery/blink/main.cpp` should be a newline value. Maybe your editor converted something? Hm, but it's weird that it's only for F0 projects.

```cpp
#include␣<xpcc/architecture/platform.hpp>¬
¬
using␣namespace␣Board;¬
¬
int¬
main()¬
{¬
>·······Board::initialize();¬
¬
>·······LedGreen::set();¬
>·······LedBlue::reset();¬
¬
>·······while␣(true)␣{¬
>·······>·······LedGreen::toggle();¬
>·······>·······LedBlue::toggle();¬
>·······>·······xpcc::delayMilliseconds(Board::Button::read()␣?␣500␣:␣1000);¬
>·······}¬
}¬
```

**Comment by lmoesch** [comment]
Jeah, despite the error message, it does not seem to have anything todo with the main.cpp. When I change the build target, the error disappears, so I guess it has some problems with parsing/loading the device files.

I even tried the following code:

`int
main(){
while(1)
{}};`

Still the same error.

Can you reproduce the error on Linux?

**Comment by lmoesch** [comment]
The root of the problem were carelessly used "io.open" calls in the template.py, wich leads to encoding errors, when pythons default encoding ist not utf-8 (which it is unfortunately not under Windows /sigh). I temporarily fixed it, by specifying the encoding directly in all io.open calls. I guess the "correct" way would be trying to get the encoding from file or atleast try-catching all io.open/read calls so that useful error messages can be displayed.

**Comment by salkinium** [comment]
Hm, this sucks, this was probably a side-effect of switching to Python 3, which IIRC has different behavior on file encodings.

**Comment by ekiwi** [comment]
Can we just always use utf-8?
@lmoesch : how did you specify the encoding for the `io.open` call? Should we do this for all code that reads a file to a string?

**Comment by lmoesch** [comment]
@salkinium I tested it with python2 aswell. This seems to be a python specific thing on Windows. (To be more specific the default encoding is **hardcoded** in the platformfile of the python installation)

@ekiwi 
`open(_path_, _mode_, encoding='utf-8')` //for all calls in template.py
You should always specify the target file's encoding, since there is no guarantee that your file has the "right" encoding. Since there is no good way to determine a files encoding beforehand, in my oppinion it is totally okay to say, that all files have to be utf-8 encoded.

**Comment by salkinium** [comment]
Fixed in #349.
